### PR TITLE
ecdsa-multikey v1.0.0

### PR DESCRIPTION
**Pull request by kezike**
Version 1.0.0 of `ecdsa-multikey`:
- Implementation of ECDSA key pair generation with curves `P-256`, `P-384`, and `P-521`
- Implementation of ECDSA sign and verify methods
- Tests for ECDSA key pair generation with curves `P-256`, `P-384`, and `P-521`
- Tests for ECDSA sign and verify methods

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
+name: Node.js CI
+
+on: [push]
+
+jobs:
+  test-node:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        node-version: [16.x]
```
Comment (davidlehn) [inline .github/workflows/main.yml:10]: Test on 18.x too:
```suggestion
        node-version: [16.x, 18.x]
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    steps:
+    - uses: actions/checkout@v2
```
Comment (davidlehn) [inline .github/workflows/main.yml:12]: Change all `actions/checkout` and `actions/setup-node` to `@3`.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    - name: Use Node.js ${{ matrix.node-version }}
+      uses: actions/setup-node@v1
+      with:
+        node-version: ${{ matrix.node-version }}
+    - run: npm install
+    - name: Generate coverage report
+      run: npm run coverage-ci
+    - name: Upload coverage to Codecov
+      uses: codecov/codecov-action@v2
```
Comment (davidlehn) [inline .github/workflows/main.yml:64]:
```suggestion
      uses: codecov/codecov-action@v3
```

File: karma.conf.cjs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    frameworks: ['mocha', 'chai'],
+
+    // list of files / patterns to load in the browser
+    files: [
+      'test/*.spec.js'
+    ],
+
+    // list of files to exclude
+    exclude: [
+      'test/compatibility.spec.js'
```
Comment (davidlehn) [inline karma.conf.cjs:21]: I don't see this file.  Is this needed?

File: lib/factory.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
+/*!
+ * Copyright (c) 2023 Digital Bazaar, Inc. All rights reserved.
+ */
+import {ALGORITHM} from './constants.js';
+import {webcrypto, EcdsaCurve, EcdsaHash} from './ecdsa.js';
+
+export function createSigner({id, secretKey}) {
+  if(!secretKey) {
+    throw new Error('A secret key is not available for signing.');
```
Comment (davidlehn) [inline lib/factory.js:9]: Not sure about the phrasing here and in verify call below.  Maybe it should be more direct such as "secretKey required for signing."?   I'm not sure what we use in other libraries that are not just using an assert library.  Also could be `TypeError`s.

File: lib/ecdsa.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+/*!
+ * Copyright (c) 2023 Digital Bazaar, Inc. All rights reserved.
+ */
+import {webcrypto} from 'node:crypto';
+const {CryptoKey} = webcrypto;
+export {CryptoKey, webcrypto};
+
+export const EcdsaCurve = {
```
Comment (davidlehn) [inline lib/ecdsa.js:8]: Do `EcdsaCurve` and `EcdsaHash` need to be somewhere else so they exist when `ecdsa-browser.js` is in use?  Maybe the constants file?  If they move there, then these files could be renamed `crypto{-browser}.js`, or similar, since that's all that would contain.

File: lib/helpers.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      return EcdsaCurve.P384;
+    }
+    if(publicMultikey[1] === MULTICODEC_P521_PUBLIC_KEY_HEADER_BYTE_2) {
+      return EcdsaCurve.P521;
+    }
+  }
+
+  // FIXME; also support P-256K/secp256k1
+  const err = new Error('Unsupported multikey header.');
+  err.name = 'UnsupportedError';
```
Comment (davidlehn) [inline lib/helpers.js:33]: Why is this explicitly named here, but not in the below "unsupported" errors?

File: package.json [hunk 8]
```diff
@@ -0,0 +1,64 @@
+{
+  "name": "@digitalbazaar/ecdsa-multikey",
+  "version": "1.0.0",
```
Comment (davidlehn) [inline package.json:3]: Set to pre-1.0.0 so our release tooling increments it properly.
```suggestion
  "version": "0.0.1",
```

File: package.json [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  "browser": {
+    "./lib/ecdsa.js": "./lib/ecdsa-browser.js"
+  },
+  "files": [
+    "lib/**/*.js"
+  ],
+  "dependencies": {
+    "base58-universal": "^2.0.0",
+    "base64url-universal": "^2.0.0",
+    "multiformats": "^11.0.1"
```
Comment (davidlehn) [inline package.json:22]: Used?

File: lib/index.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+  return keyPair;
+}
+
+function _assertMultikey(key) {
+  if(!(key && typeof key === 'object')) {
+    throw new TypeError('"key" must be an object.');
+  }
+  if(key.type !== 'Multikey') {
+    throw new Error('"key" must be a Multikey with type "Multikey".');
```
Comment (davidlehn) [inline lib/index.js:89]:
```suggestion
    throw new TypeError('"key" must be a Multikey with type "Multikey".');
```

File: lib/serialize.js [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+  if(!publicKeyMultibase) {
+    throw new TypeError('The "publicKeyMultibase" property is required.');
+  }
+
+  const keyPair = {id};
+
+  // import public key
+  if(!(publicKeyMultibase && typeof publicKeyMultibase === 'string' &&
+    publicKeyMultibase[0] === MULTIBASE_BASE58_HEADER)) {
+    throw new Error(
```
Comment (davidlehn) [inline lib/serialize.js:129]:
```suggestion
    throw new TypeError(
```

File: lib/serialize.js [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+  // compressed public keys
+  const spki = _toSpki({publicMultikey});
+  keyPair.publicKey = await webcrypto.subtle.importKey(
+    'spki', spki, algorithm, EXTRACTABLE, ['verify']);
+
+  // import secret key if given
+  if(secretKeyMultibase) {
+    if(!(typeof secretKeyMultibase === 'string' &&
+    secretKeyMultibase[0] === MULTIBASE_BASE58_HEADER)) {
+      throw new Error(
```
Comment (davidlehn) [inline lib/serialize.js:150]:
```suggestion
      throw new TypeError(
```

**Review by davidlehn — COMMENTED** [review]
Please also fix issues from `npm run lint`.

File: lib/factory.js (hunk 5 above)
Comment (kezike) [reply lib/factory.js:10]: This error message was used in the `ed25519-multikey` library, so I decided to add it here for consistency. @gannan08 thoughts?

File: lib/ecdsa.js (hunk 6 above)
Comment (kezike) [reply lib/ecdsa.js:9]: Yes, they most likely do because I noticed that this caused errors in the karma test. I will fix this soon.

File: package.json (hunk 9 above)
Comment (kezike) [reply package.json]: At one point I was using it, but not anymore. I have removed it for now, but we may need it in the future, since `multicodec` is deprecated and will not support future additions to the `multiformats/multicodec` table.

File: lib/factory.js (hunk 5 above)
Comment (kezike) [reply lib/factory.js:10]: For now, I will update these to more actionable error messages.

File: lib/translators.js [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+} from './constants.js';
+
+// converts key pair to Multikey format
+export async function toMultikey({keyPair}) {
+  const validEcdsaTypes = [
+    ECDSA_2019_SECP_256_KEY_TYPE,
+    ECDSA_2019_SECP_384_KEY_TYPE,
+    ECDSA_2019_SECP_521_KEY_TYPE
+  ];
+  if(!validEcdsaTypes.includes(keyPair.type)) {
```
Comment (kezike) [inline lib/translators.js:19]: @gannan08 Is this check necessary/proper? I found these key type constants in the ECDSA 2019 Data Integrity spec (e.g., [`EcdsaSecp256r1VerificationKey2019`](https://w3c-ccg.github.io/di-ecdsa-2019/#ecdsasecp256r1verificationkey2019)) and I saw a similar check in the ED25519 Multikey library [here](https://github.com/digitalbazaar/ed25519-multikey/blob/main/lib/keyPairTranslator.js#L8). However, this may not be a necessary check for this special, especially since these constants that are defined in the spec are for SECP keys and may not even be necessary with the `Multikey` type.

File: README.md [hunk 14]
```diff
@@ -82,9 +82,9 @@ await keyPair.export({publicKey: true});
 // ->
 {
   type: 'Multikey',
-  id: 'did:example:1234#z6Mkon3Necd6NkkyfoGoHxid2znGc59LU3K7mubaRcFbLfLX',
+  id: 'did:example:1234#zynkWQ2LUf58H4VwLGjscwb9KJwGUmqBFXdJKnCXxDMciFXioY7Hq7MvEdVfsBiQSVa9k9',
```
Comment (gannan08) [inline README.md:85]: Ensure that the fingerprint corresponds to a generated keypair. The header for the curves should be the following (as of [962df2013b](https://github.com/digitalbazaar/ecdsa-multikey/tree/962df2013baec09d39fa66fa894436edf1b027ec)):


```
"zDnae..." => P-256
"z82Lk..." => P-384
"z2J9g..." => P-521
```

File: README.md [hunk 15]
```diff
@@ -98,10 +98,10 @@ await keyPair.export({publicKey: true, secretKey: true});
...
   controller: 'did:example:1234',
-  publicKeyMultibase: 'z6Mkon3Necd6NkkyfoGoHxid2znGc59LU3K7mubaRcFbLfLX',
-  secretKeyMultibase: 'zruzf4Y29hDp7vLoV3NWzuymGMTtJcQfttAWzESod4wV2fbPvEp4XtzGp2VWwQSQAXMxDyqrnVurYg2sBiqiu1FHDDM'
+  publicKeyMultibase: 'zynkWQ2LUf58H4VwLGjscwb9KJwGUmqBFXdJKnCXxDMciFXioY7Hq7MvEdVfsBiQSVa9k9',
+  secretKeyMultibase: 'zEbCcd62anxNxioGCdP818UobJn1oEmgL5PHUvQ5AvdyVeLJxeJ4Kr73RqqvG5DN7Zq48'
```
Comment (gannan08) [inline README.md:104]: Ensure proper encoding of `publicKeyMultibase` and  `secretKeyMultibase`.

File: lib/constants.js [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+// ECDSA curves
+export const ECDSA_CURVE = {
+  P256: 'P-256',
+  P384: 'P-384',
+  P521: 'P-521'
+};
+
+// ECDSA hash functions
+export const ECDSA_HASH = {
+  Sha256: 'SHA-256',
```
Comment (gannan08) [inline lib/constants.js:45]:
```suggestion
  SHA256: 'SHA-256',
```

File: lib/constants.js [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+  Sha384: 'SHA-384',
```
Comment (gannan08) [inline lib/constants.js:46]:
```suggestion
  SHA384: 'SHA-384',
```

File: lib/constants.js [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+  Sha512: 'SHA-512'
```
Comment (gannan08) [inline lib/constants.js:47]:
```suggestion
  SHA512: 'SHA-512'
```

File: lib/helpers.js [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+    if(publicMultikey[1] === _getLastByte(MULTICODEC_P521_PUBLIC_KEY_HEADER)) {
+      return ECDSA_CURVE.P521;
+    }
+  }
+
+  // FIXME; also support P-256K/secp256k1
+  throw new TypeError('Unsupported multikey header.');
+}
+
+// retrieves size of secret key
```
Comment (gannan08) [inline lib/helpers.js:36]:
```suggestion
// retrieves byte size of secret key
```

File: lib/serialize.js [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+} from './constants.js';
+import {webcrypto} from './crypto.js';
+import {
+  getNamedCurve,
+  getSecretKeySize,
+  setPublicKeyHeader,
+  setSecretKeyHeader
+} from './helpers.js';
+
+// FIXME: may need to move any leading zeros for bitstring compression; needs
```
Comment (gannan08) [inline lib/serialize.js:21]: We will need to address this FIXME before releasing v1. This should be completed in a separate PR.

cc: @dlongley

File: lib/serialize.js [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+  // import secret key if given
+  if(secretKeyMultibase) {
+    if(!(typeof secretKeyMultibase === 'string' &&
+    secretKeyMultibase[0] === MULTIBASE_BASE58_HEADER)) {
+      throw new TypeError(
+        '"secretKeyMultibase" must be a multibase, base58-encoded string.');
+    }
+    const secretMultikey = base58.decode(secretKeyMultibase.slice(1));
+
+    // FIXME: ensure secret key multikey header appropriately matches the
```
Comment (gannan08) [inline lib/serialize.js:161]: @dlongley This doesn't make sense to me. What is a "multikey header"? Perhaps, this is in reference to the `publicKeyMultibase` header on the Multikey. Did you mean ensure that the `publicKeyMultibase` matches its expected corresponding  `secretKeyMultibase`?

@kezike We will need to address this in another PR before release.

File: package.json [hunk 22]
```diff
@@ -10,12 +10,16 @@
...
+    "./lib/crypto.js": "./lib/crypto-browser.js"
+  },
   "files": [
     "lib/**/*.js"
   ],
   "dependencies": {
-    "base58-universal": "^2.0.0"
+    "base58-universal": "^2.0.0",
+    "base64url-universal": "^2.0.0",
+    "varint": "^6.0.0"
```
Comment (gannan08) [inline package.json:22]: Move to `devDependencies`. This is not used at runtime.

File: package.json [hunk 23]
```diff
@@ -35,7 +39,7 @@
     "mocha": "^10.0.0",
     "mocha-lcov-reporter": "^1.3.0",
     "multibase": "^3.1.0",
-    "multicodec": "^2.0.1",
+    "multicodec": "^3.2.1",
```
Comment (gannan08) [inline package.json:42]: We will need to move off of this package since it's deprecated and will not receive the new p*-priv multicodec codes. This change should be addressed in a separate PR and completed before release.

**Review by gannan08 — CHANGES_REQUESTED** [review]
Please create issues on this repo for the FIXMEs that will need to be addressed in a separate PR. There are a few minor nits. Also, requested some clarification from @dlongley.

We're almost there.

File: lib/factory.js (hunk 5 above)
Comment (gannan08) [reply lib/factory.js]: @kezike +1. That was going to be my recommendation.

File: lib/translators.js [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+  ECDSA_2019_SECP_256_KEY_TYPE,
+  ECDSA_2019_SECP_384_KEY_TYPE,
+  ECDSA_2019_SECP_521_KEY_TYPE,
+  ECDSA_2019_SUITE_CONTEXT_V1_URL,
+  MULTIKEY_CONTEXT_V1_URL
+} from './constants.js';
+
+// converts key pair to Multikey format
+export async function toMultikey({keyPair}) {
+  const validEcdsaTypes = [
```
Comment (gannan08) [inline lib/translators.js:14]: Move out of the function and into a constant in the file. The values in the array are constant between function invocations, therefore we don't need to create a new array on each function invocation.

Also, change the array to a set after moving the code.

File: lib/serialize.js (hunk 20 above)
Comment (kezike) [reply lib/serialize.js:21]: Created Issue #2 to address this `FIXME` comment

File: lib/serialize.js (hunk 21 above)
Comment (kezike) [reply lib/serialize.js:161]: Created Issue #3 to address this `FIXME` comment

File: package.json (hunk 23 above)
Comment (kezike) [reply package.json:41]: Created Issue #4 to address this `FIXME` comment

File: lib/crypto-browser.js [hunk 25]
```diff
@@ -0,0 +1,7 @@
+/*!
+ * Copyright (c) 2023 Digital Bazaar, Inc. All rights reserved.
+ */
+// eslint-disable-next-line no-undef
+export const webcrypto = globalThis.crypto.webcrypto ?? globalThis.crypto;
```
Comment (dlongley) [inline lib/crypto-browser.js:5]: Do any browsers actually do `globalThis.crypto.webcrypto`? I thought this was node.js-specific. If we don't know of any browsers that do this, we should remove it. I would think that would also mean we could simplify the following line too.

File: lib/factory.js [hunk 26]
```diff
@@ -0,0 +1,51 @@
+/*!
+ * Copyright (c) 2023 Digital Bazaar, Inc. All rights reserved.
```
Comment (dlongley) [inline lib/factory.js:2]: This stuff was copied over from work done last year, so we should include that in the copyright.

```suggestion
 * Copyright (c) 2022-2023 Digital Bazaar, Inc. All rights reserved.
```

File: lib/serialize.js [hunk 27]
```diff
@@ -0,0 +1,230 @@
+/*!
+ * Copyright (c) 2023 Digital Bazaar, Inc. All rights reserved.
```
Comment (dlongley) [inline lib/serialize.js:2]:
```suggestion
 * Copyright (c) 2022-2023 Digital Bazaar, Inc. All rights reserved.
```

File: lib/helpers.js [hunk 28]
```diff
@@ -0,0 +1,96 @@
+/*!
+ * Copyright (c) 2023 Digital Bazaar, Inc. All rights reserved.
```
Comment (dlongley) [inline lib/helpers.js:2]:
```suggestion
 * Copyright (c) 2022-2023 Digital Bazaar, Inc. All rights reserved.
```

File: lib/constants.js [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+// Multikey context v1 URL
+export const MULTIKEY_CONTEXT_V1_URL = 'https://w3id.org/security/multikey/v1';
+export const MULTIBASE_BASE58_HEADER = 'z';
+
+// Multicodec p256-pub header (0x1200 varint -> 0x8024 hex)
+export const MULTICODEC_P256_PUBLIC_KEY_HEADER = 0x8024;
+// Multicodec p384-pub header (0x1201 varint -> 0x8124 hex)
+export const MULTICODEC_P384_PUBLIC_KEY_HEADER = 0x8124;
+// Multicodec p521-pub header (0x1202 varint -> 0x8224 hex)
+export const MULTICODEC_P521_PUBLIC_KEY_HEADER = 0x8224;
```
Comment (dlongley) [inline lib/constants.js:27]: We should change these to use a Uint8Array and then remove the extra getFirstByte/getLastByte helpers to simplify:

```suggestion
export const MULTICODEC_P256_PUBLIC_KEY_HEADER =
  new Uint8Array([0x80, 0x24]);
// Multicodec p384-pub header (0x1201 varint -> 0x8124 hex)
export const MULTICODEC_P384_PUBLIC_KEY_HEADER =
  new Uint8Array([0x81, 0x24]);
// Multicodec p521-pub header (0x1202 varint -> 0x8224 hex)
export const MULTICODEC_P521_PUBLIC_KEY_HEADER
  new Uint8Array([0x82, 0x24]);
```

File: lib/constants.js [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+
+// Multicodec p256-priv header (0x1306 varint -> 0x8626 hex)
+export const MULTICODEC_P256_SECRET_KEY_HEADER = 0x8626;
+// Multicodec p384-priv header (0x1307 varint -> 0x8726 hex)
+export const MULTICODEC_P384_SECRET_KEY_HEADER = 0x8726;
+// Multicodec p521-priv header (0x1308 varint -> 0x8826 hex)
+export const MULTICODEC_P521_SECRET_KEY_HEADER = 0x8826;
```
Comment (dlongley) [inline lib/constants.js:34]:
```suggestion
export const MULTICODEC_P256_SECRET_KEY_HEADER =
  new Uint8Array([0x86, 0x26]);
// Multicodec p384-priv header (0x1307 varint -> 0x8726 hex)
export const MULTICODEC_P384_SECRET_KEY_HEADER =
  new Uint8Array([0x87, 0x26]);
// Multicodec p521-priv header (0x1308 varint -> 0x8826 hex)
export const MULTICODEC_P521_SECRET_KEY_HEADER =
  new Uint8Array([0x88, 0x26]);
```

File: lib/helpers.js [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+  if(publicMultikey[0] === _getFirstByte(MULTICODEC_P384_PUBLIC_KEY_HEADER)) {
+    if(publicMultikey[1] === _getLastByte(MULTICODEC_P384_PUBLIC_KEY_HEADER)) {
+      return ECDSA_CURVE.P384;
+    }
+  }
+  if(publicMultikey[0] === _getFirstByte(MULTICODEC_P521_PUBLIC_KEY_HEADER)) {
+    if(publicMultikey[1] === _getLastByte(MULTICODEC_P521_PUBLIC_KEY_HEADER)) {
+      return ECDSA_CURVE.P521;
+    }
+  }
```
Comment (dlongley) [inline lib/helpers.js:30]:
```suggestion
  if(publicMultikey[0] === MULTICODEC_P256_PUBLIC_KEY_HEADER[0] &&
    publicMultikey[1] === MULTICODEC_P256_PUBLIC_KEY_HEADER[1]) {
    return ECDSA_CURVE.P256;
  }
  if(publicMultikey[0] === MULTICODEC_P384_PUBLIC_KEY_HEADER[0] &&
    publicMultikey[1] === MULTICODEC_P384_PUBLIC_KEY_HEADER[1]) {
    return ECDSA_CURVE.P384;
  }
  if(publicMultikey[0] === MULTICODEC_P521_PUBLIC_KEY_HEADER[0] &&
    publicMultikey[1] === MULTICODEC_P521_PUBLIC_KEY_HEADER[1]) {
    return ECDSA_CURVE.P521;
  }
```

File: lib/helpers.js [hunk 32]
```diff
@@ -0,0 +1,96 @@
...
+    return 66;
+  }
+  throw new TypeError(`Unsupported curve "${curve}".`);
+}
+
+// sets secret key header bytes on key pair
+export function setSecretKeyHeader({keyPair, buffer}) {
+  const key = keyPair.secretKey || keyPair.publicKey;
+  const {namedCurve: curve} = key.algorithm;
+  // FIXME: these must be added to the multicodec table
```
Comment (dlongley) [inline lib/helpers.js:56]: Looks like this got done (thanks!):

```suggestion
```

File: lib/helpers.js [hunk 33]
```diff
@@ -0,0 +1,96 @@
...
+  if(curve === ECDSA_CURVE.P256) {
+    buffer[0] = _getFirstByte(MULTICODEC_P256_SECRET_KEY_HEADER);
+    buffer[1] = _getLastByte(MULTICODEC_P256_SECRET_KEY_HEADER);
+  } else if(curve === ECDSA_CURVE.P384) {
+    buffer[0] = _getFirstByte(MULTICODEC_P384_SECRET_KEY_HEADER);
+    buffer[1] = _getLastByte(MULTICODEC_P384_SECRET_KEY_HEADER);
+  } else if(curve === ECDSA_CURVE.P521) {
+    buffer[0] = _getFirstByte(MULTICODEC_P521_SECRET_KEY_HEADER);
+    buffer[1] = _getLastByte(MULTICODEC_P521_SECRET_KEY_HEADER);
```
Comment (dlongley) [inline lib/helpers.js:65]:
```suggestion
  if(curve === ECDSA_CURVE.P256) {
    buffer[0] = MULTICODEC_P256_SECRET_KEY_HEADER[0];
    buffer[1] = MULTICODEC_P256_SECRET_KEY_HEADER[1];
  } else if(curve === ECDSA_CURVE.P384) {
    buffer[0] = MULTICODEC_P384_SECRET_KEY_HEADER[0];
    buffer[1] = MULTICODEC_P384_SECRET_KEY_HEADER[1];
  } else if(curve === ECDSA_CURVE.P521) {
    buffer[0] = MULTICODEC_P521_SECRET_KEY_HEADER[0];
    buffer[1] = MULTICODEC_P521_SECRET_KEY_HEADER[1];
```

File: lib/helpers.js [hunk 34]
```diff
@@ -0,0 +1,96 @@
...
+
+// get first byte of data
+function _getFirstByte(data) {
+  return data >> 8;
+}
+
+// get last byte of data
+function _getLastByte(data) {
+  return data & 0x00ff;
+}
```
Comment (dlongley) [inline lib/helpers.js:96]:
```suggestion
```

File: lib/index.js [hunk 35]
```diff
@@ -1,30 +1,103 @@
...
+import {createSigner, createVerifier} from './factory.js';
+import {exportKeyPair, importKeyPair} from './serialize.js';
+import {toMultikey} from './translators.js';
 
+// FIXME: support `P-256K` via `@noble/secp256k1`
+// generates ECDSA key pair
+export async function generate({id, controller, curve} = {}) {
+  if(!curve) {
+    throw new TypeError('"curve" must be one of the following values: ' +
+      `${Object.values(ECDSA_CURVE).map(v => `'${v}'`).join(', ')}.`);
```
Comment (dlongley) [inline lib/index.js:20]: Style nit:

```suggestion
    throw new TypeError(
      '"curve" must be one of the following values: ' +
      `${Object.values(ECDSA_CURVE).map(v => `'${v}'`).join(', ')}.`);
```

**Review by dlongley — CHANGES_REQUESTED** [review]
Thanks, this is looking really good! I just have a few simplifications / nits to apply.

File: .github/workflows/main.yml [hunk 36]
```diff
@@ -21,11 +21,11 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        node-version: [16.x]
+        node-version: [16.x, 18.x]
```
Comment (davidlehn) [inline .github/workflows/main.yml:24]: Karma is doing the testing, so just need to run it once with one node version.
```suggestion
        node-version: [18.x]
```

File: .github/workflows/main.yml [hunk 37]
```diff
@@ -35,11 +35,11 @@ jobs:
...
```
Comment (davidlehn) [inline .github/workflows/main.yml:38]: Only need to run lint once.
```suggestion
        node-version: [18.x]
```

File: lib/crypto-browser.js (hunk 25 above)
Comment (kezike) [reply lib/crypto-browser.js:5]: Ah yes, this comment in the [Node.js SDK for Web Crypto API](https://nodejs.org/api/webcrypto.html#web-crypto-api) suggests the same:

> Use globalThis.crypto or require('node:crypto').webcrypto to access this module.

I will simplify this code by reducing the entire code to one line:

```suggestion
export const CryptoKey = globalThis.CryptoKey ?? globalThis.crypto.CryptoKey;
```

Comment (kezike) [reply lib/crypto-browser.js:5]: Actually, I also need to export `webcrypto`, so I will just change the first line to this:

```suggestion
export const webcrypto = globalThis.crypto;
```

File: lib/constants.js (hunk 30 above)
Comment (dlongley) [reply lib/constants.js:33]: This looks like it wasn't applied yet.

File: lib/helpers.js [hunk 38]
```diff
@@ -0,0 +1,84 @@
...
+  MULTICODEC_P384_PUBLIC_KEY_HEADER,
+  MULTICODEC_P384_SECRET_KEY_HEADER,
+  MULTICODEC_P521_PUBLIC_KEY_HEADER,
+  MULTICODEC_P521_SECRET_KEY_HEADER
+} from './constants.js';
+
+// retrieves name of appropriate ECDSA curve
+export function getNamedCurve({publicMultikey}) {
+  if(publicMultikey[0] === MULTICODEC_P256_PUBLIC_KEY_HEADER[0]) {
+    if(publicMultikey[1] === MULTICODEC_P256_PUBLIC_KEY_HEADER[1]) {
```
Comment (dlongley) [inline lib/helpers.js:17]: These can all be made into compound AND statements:

```suggestion
  if(publicMultikey[0] === MULTICODEC_P256_PUBLIC_KEY_HEADER[0] &&
    publicMultikey[1] === MULTICODEC_P256_PUBLIC_KEY_HEADER[1]) {
```

**Review by dlongley — CHANGES_REQUESTED** [review]
Looks like a couple of suggestions were missed from before, but then we should be good to go!

**Review by dlongley — APPROVED** [review]
Thanks! @gannan08 let me know if you need me to merge and release this or if you will.

**Comment by gannan08** [comment]
@dlongley Can you please merge and release.
